<!DOCTYPE HTML>
<html>
<head>
<title>Try</title>
<meta name="keywords" content="Befehl, Ablaufsteuerung, Fehler">
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<link href="../static/theme.css" rel="stylesheet" type="text/css" />
<link href="../static/font.css" rel="stylesheet" type="text/css" />
<script src="../static/ga.js" type="text/javascript"></script>
<!--[if lt IE 9]>
<script src="../static/html5.js"></script>
<![endif]-->
<script src="../navbar/NavScript.js" type="text/javascript"></script>
</head>
<body>
<script type="text/javascript">WriteHeader(1);</script>
<div id="content">
  <section>
    <div id="article">
      <h1>Try <span class="ver">[v1.1.04+]</span></h1>
      <p>Schützt eine oder mehrere Anweisungen (Befehle oder Ausdrücke) vor Laufzeitfehlern und Ausnahmen, die vom <a href="Throw.htm">Throw</a>-Befehl ausgelöst werden.</p>
      <pre class="Syntax">Try <i>Anweisung</i></pre>
      <pre class="Syntax" style="line-height: 100%">Try
{
    <i>Anweisungen</i>
}</pre>
      <h3>Bemerkungen</h3>
      <p>Nach dem <em>Try</em>-Befehl erfolgt üblicherweise ein <a href="Block.htm">Block</a> - das sind eine oder mehrere Anweisungen (Befehle oder Ausdrücke), umschlossen von geschweiften Klammern. Verwendet man nur eine Anweisung, kann sie direkt neben <em>Try</em> gesetzt werden, oder darunter ohne geschweifte Klammern. Sollen Zeilen nur ausgeführt werden, wenn <em>Try</em> einen Fehler abfängt, verwende den <a href="Catch.htm">Catch</a>-Befehl.</p>
      <p>Sobald der <a href="Throw.htm">Throw</a>-Befehl verwendet wird oder ein Laufzeitfehler beim Ausführen eines <em>Try</em>-Blocks erkannt wird, wird die Ausführung den Block verlassen. Ist eine <a href="Catch.htm">Catch</a>-Anweisung vorhanden, wird sie ausgeführt. Werden beim Auslösen einer Ausnahme keine <em>Try</em>-Blocks ausgeführt, wird eine Fehlermeldung angezeigt und der aktuelle Thread beendet.</p>
      <p>Wenn mehrere <em>Try</em>-Befehle ausgeführt werden, wird der aktuellste gerade ausführende Befehl Vorrang haben.</p>
      <p>Bei Bedarf kann der "<a href="Block.htm#otb">One True Brace Style</a>" beim <em>Try</em>-Befehl angewendet werden. Zum Beispiel:</p>
      <pre>try <strong>{</strong>
    ...
<strong>}</strong> catch e <strong>{</strong>
    ...
<strong>}</strong></pre>
      <h3>Verwandte Befehle</h3>
      <p><a href="Catch.htm">Catch</a>, <a href="Throw.htm">Throw</a>, <a href="Finally.htm">Finally</a>, <a href="Block.htm">Blocks</a></p>
      <h3 id="Examples">Beispiele</h3>
      <pre class="NoIndent" id="ex_basic"><em>; Beispiel #1: Das Grundkonzept von try/catch/throw.</em>

try  <em>; Versucht, den Code auszuführen.</em>
{
    HelloWorld()
    MakeToast()
}
<a href="Catch.htm">catch</a> e  <em>; Verarbeitet den ersten Fehler/Ausnahme, ausgelöst durch den vorherigen Block.</em>
{
    MsgBox, Eine Ausnahme wurde ausgelöst!`nSpeziell: %e%
    <a href="Exit.htm">Exit</a>
}

HelloWorld()  <em>; Ist immer erfolgreich.</em>
{
    MsgBox, Hello, world!
}

MakeToast()  <em>; Schlägt immer fehl.</em>
{
    <em>; Springt sofort zur Fehlerbehandlung des Try-Blocks:</em>
    <a href="Throw.htm">throw</a> A_ThisFunc " ist nicht implementiert, sorry"
}
</pre>
      <pre class="NoIndent" id="ex_el"><em>; Beispiel #2: Verwendet try/catch anstelle von ErrorLevel.</em>

try
{
    <em>; Nachfolgend wird versucht, bestimmte Dateitypen abzusichern:</em>
    FileCopy, %A_MyDocuments%\*.txt, D:\Backup\Textdokumente
    FileCopy, %A_MyDocuments%\*.doc, D:\Backup\Textdokumente
    FileCopy, %A_MyDocuments%\*.jpg, D:\Backup\Fotos
}
catch
{
    MsgBox, 16,, Es trat ein Fehler beim Absichern der Dateien auf!
    ExitApp
}
</pre>
      <pre class="NoIndent" id="ex_com"><em>; Beispiel #3: Umgang mit COM-Fehlern.</em>

try
{
    obj := <a href="ComObjCreate.htm">ComObjCreate</a>("<a href="http://msdn.microsoft.com/en-us/library/aa227633(v=vs.60).aspx">ScriptControl</a>")
    obj.ExecuteStatement("MsgBox ""Das ist ein eingebettetes VBScript""")
    obj.InvalidMethod() <em>; Diese Zeile erzeugt einen Laufzeitfehler.</em>
}
catch e
{
    <em>; Um mehr über das Objekt zu erfahren, das in e enthalten ist, siehe <a href="Catch.htm#RuntimeErrors">Catch</a>.</em>
    MsgBox, 16,, % "Ausnahme ausgelöst!`n`nWas: " e.what &quot;`nDatei: " e.file
        . &quot;`nZeile: " e.line &quot;`nMeldung: " e.message &quot;`nExtra: " e.extra
}
</pre>
      <pre class="NoIndent" id="ex_nesting"><em>; Beispiel #4: Verschachtelte Try-Catch-Anweisungen.</em>

try Beispiel1() <em>; Jede Einzelanweisung kann auf der selben Zeile des Try-Befehls stehen.</em>
catch e
    MsgBox, Beispiel1() hat "%e%" ausgelöst.

Beispiel1()
{
    try Beispiel2()
    catch e
    {
        if e = 1
            throw e <em>; Ausnahme wird nochmals ausgelöst, so dass der Aufrufer sie abfangen kann.</em>
        else
            MsgBox, Beispiel2() hat "%e%" ausgelöst.
    }
}

Beispiel2()
{
    Random, o, 1, 2
    throw o
}</pre>
    </div>
  </section>
</div>
<script type="text/javascript">WriteFooter(1);</script>
</body>
</html>